### PR TITLE
feat: bulk book users to sessions via csv file

### LINKED ISSUE #150: Be able to upload bookings (link users to a session in bulk) via csv upload

### PR DESCRIPTION
**Pull request by keevan**
Summary:
- permission check to only allow users with this permission to do so
- can intially set notification type, status (booked/cancelled) and
  discount code.
- tests for various scenarios.
- managed via a booking manager class.
- supports both csv file and array (list of bookings) formats.

Resolves #150

TODO: 
- [x] add event handling(?)

### CODE REVIEW

**Comment by keevan** [comment]
Pulled in some changes found in 403, goal is that the behat issues have already been resolved.

File: classes/booking_manager.php [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+
+            // Set defaults for fields with no value.
+            $entry->status = $entry->status ?? '';
+            $entry->notificationtype = $entry->notificationtype ?? '';
+            $entry->discountcode = $entry->discountcode ?? '';
+
+            // Validate and get user.
+            $userids = $DB->get_records('user', ['email' => $entry->email], 'id');
+
+            // Multiple matched, ambiguous which is the real one.
```
Comment (Peterburnett) [inline classes/booking_manager.php:180]: Given that these have a couple of conditions on most of the fields, I reckon we should just split this up into a couple of private functions that just take a field -> errors, or row -> errors if there is cross dependencies. that way the main loop just reads as the series of checks being done at a high level

File: classes/form/confirm_bookings_form.php [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        $mform->addElement('hidden', 'f');
+        $mform->setType('f', PARAM_INT);
+
+        // Reference to the uploaded file.
+        $mform->addElement('hidden', 'fileid', $fileid);
+        $mform->setType('fileid', PARAM_INT);
+
+        $backurl = new moodle_url('/mod/facetoface/upload.php', ['f' => $f]);
+        $htmlbuttons = $OUTPUT->render((new single_button(
+            new moodle_url('upload.php', ['f' => $f, 'fileid' => $fileid, 'process' => 1]),
```
Comment (Peterburnett) [inline classes/form/confirm_bookings_form.php]: one is an absolute URL, one is a relative, make these consistent?

File: classes/form/upload_bookings_form.php [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+ *
+ * @package    mod_facetoface
+ * @author     Kevin Pham <kevinpham@catalyst-au.net>
+ * @copyright  Catalyst IT, 2024
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class upload_bookings_form extends \moodleform {
+
+    /**
+     * Build form for importing woekflows.
```
Comment (Peterburnett) [inline classes/form/upload_bookings_form.php]: typo in comment

File: classes/booking_manager.php (hunk 1 above)
Comment (keevan) [reply classes/booking_manager.php:180]: I tested for user, and that was okay.., being that it worked when I added a method like `check_user($email, &$errors, $row) -> $userid`, but the problem is it introduced some level of inconsistency in how the checks are performed and recorded compared to sessions, which required different parameters to be provided, along with a session cache object, so it's probably better to keep all the checks inline for now, until there's a better path forward for all the checks collectively.

File: classes/form/upload_bookings_form.php [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+     * @see \moodleform::definition()
+     */
+    public function definition() {
+        global $CFG;
+
+        $mform = $this->_form;
+
+        $mform->addElement('header', 'settingsheader', get_string('upload'));
+
+        $url = new moodle_url('example.csv');
```
Comment (Peterburnett) [inline classes/form/upload_bookings_form.php]: Not sure if there is a convention or code style guide on this, but generally im partial to absolute URLs where we can, so the PHP file/function can be moved around without invisibly breaking a link to a static asset

Comment (keevan) [reply classes/form/upload_bookings_form.php]: I blatantly copied the style from here:
https://github.com/moodle/moodle/blob/master/admin/tool/uploaduser/user_form.php#L43

But I can change it to absolute if preferred.

File: upload.php [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+    throw new moodle_exception('error:incorrectfacetofaceid', 'facetoface');
+}
+if (!$course = $DB->get_record('course', ['id' => $facetoface->course])) {
+    throw new moodle_exception('error:coursemisconfigured', 'facetoface');
+}
+if (!$cm = get_coursemodule_from_instance('facetoface', $facetoface->id, $course->id)) {
+    throw new moodle_exception('error:incorrectcoursemoduleid', 'facetoface');
+}
+
+
```
Comment (Peterburnett) [inline upload.php]: double newline here

File: upload.php [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+require_course_login($course, true, $cm);
+$context = context_course::instance($course->id);
+$modulecontext = context_module::instance($cm->id);
+require_capability('mod/facetoface:editsessions', $context);
+require_capability('mod/facetoface:uploadbookings', $context);
+
```
Comment (Peterburnett) [inline upload.php:56]: another one

File: upload.php [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+    $bm->load_from_file($fileid);
+
+    // Validate entries.
+    $errors = $bm->validate();
+    if (empty($errors)) {
+        // Process entries.
+        $bm->process();
+
+        echo "PROCESSED";
+        die;
```
Comment (Peterburnett) [inline upload.php]: leftover die?

File: upload.php [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+        // Redirect back to start with notification.
+        redirect(
+            new moodle_url('/mod/facetoface/upload.php', ['f' => $f]),
+            'Successfully processed X records.',
```
Comment (Peterburnett) [inline upload.php]: This bit looks like it wants to be a parameterised lang string?

File: upload.php (hunk 5 above)
Comment (keevan) [reply upload.php]: Good catch

File: upload.php [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+
+    $errmsg = get_string('error:bookingsuploadfileerrorsfound', 'mod_facetoface', count($errors));
+    redirect(
+        new moodle_url('/mod/facetoface/upload.php', ['f' => $f]),
+        $errmsg,
+        null,
+        notification::NOTIFY_ERROR);
+} else {
+    $mform = new upload_bookings_form(null);
+    $data = $mform->get_data();
```
Comment (Peterburnett) [inline upload.php]: is this get_data call actually needed?

File: upload.php (hunk 7 above)
Comment (keevan) [reply upload.php]: Whoops.

File: upload.php (hunk 8 above)
Comment (keevan) [reply upload.php]: I made it a bit more general. The number of records processed would equal the number of rows in the file. I don't think that information is important as it's either all or nothing, as it wouldn't redirect if there was an exception during processing.

File: classes/booking_manager.php (hunk 1 above)
Comment (keevan) [reply classes/booking_manager.php:180]: For now I've left it inline and added a comment to the function (which was missing at the time)

**Comment by danmarsden** [comment]
I'm a bit confused here - is this a new feature? - or just a backport? 

new features should land in the latest branch first - then we can deal with the backports?

**Comment by keevan** [comment]
> I'm a bit confused here - is this a new feature? - or just a backport?
> 
> new features should land in the latest branch first - then we can deal with the backports?

Good catch, I'll re-target the latest branch and create a backport PR once it's merged. Thanks Dan.

**Review by Peterburnett — APPROVED** [review]
I've largely reviewed the major commit, rather than the CI changes, but this one is looking good to go.

**Comment by Peterburnett** [comment]
@danmarsden I've approved this one, its fairly self contained, did you have anything you wanted to mention before merge?

**Comment by danmarsden** [comment]
if you've reviewed the full patch...and hard-coded english lang strings are fixed like this one I spotted:
https://github.com/catalyst/moodle-mod_facetoface/pull/151/files#diff-2359f56e8ab786049c0f9961df9bd8363b0f1e4e986972ebc692302ebc920000R119

then feel free to merge. :-)

File: db/access.php [hunk 10]
```diff
@@ -30,160 +30,170 @@
...
             'editingteacher' => CAP_ALLOW,
-            'manager' => CAP_ALLOW
-        ]
-    ]
-);
+            'manager' => CAP_ALLOW,
+        ],
+    ],
+
+    // Ability to add a new facetoface activity to a course.
```
Comment (Peterburnett) [inline db/access.php]: Comment needs updating

Comment (keevan) [reply db/access.php]: Good spot. Fixed

**Review by Peterburnett — APPROVED** [review]

**Pull request merged**